File: MCL_by_python/BeamModel.py
import CommonFunctions
import numpy as np
import WeighingFactors
import math

s_Sqrt2PI = math.sqrt(2 * math.pi)
global LambdaShort, MaxRange, m_MeasurementVariance, m_MeasurementSigma
m_MeasurementVariance, m_MeasurementSigma =0,0

def getMeasurementVariance():
    global m_MeasurementVariance
    return m_MeasurementVariance


def setMeasurementVariance(value):
    global m_MeasurementVariance
    if value != m_MeasurementVariance:
        m_MeasurementVariance = value
        sigma = math.sqrt(m_MeasurementVariance)
        m_NormalDistributionFactor = 1.0 / s_Sqrt2PI / sigma
        MeasurementSigma_set(sigma)
        # OnPropertyChanged("MeasurementVariance");
    else:
        pass


def MeasurementSigma_get():
    global m_MeasurementSigma
    return m_MeasurementSigma


def MeasurementSigma_set(value):
    global m_MeasurementVariance,  m_MeasurementSigma
    if value != m_MeasurementVariance:
        m_MeasurementSigma = value
        setMeasurementVariance (m_MeasurementSigma * m_MeasurementSigma)
    else:
        pass


def GetPHit(measuredDistance, realDistance, deltaR):
    if realDistance > MaxRange:
        return 1 - CommonFunctions.DistributionFunction(MaxRange, MaxRange, MaxRange)
    lower = CommonFunctions.DistributionFunction((measuredDistance - deltaR) * 1.0, realDistance * 1.0, getMeasurementVariance())
    upper = CommonFunctions.DistributionFunction((measuredDistance + deltaR) * 1.0, realDistance * 1.0, getMeasurementVariance())
    return upper - lower


def GetPShort(measuredDistance, realDistance, deltaR):
    if measuredDistance > realDistance:
        return 0.0
    else:
        lower = 1 - math.exp(-LambdaShort * (measuredDistance - deltaR))
        upper = 1 - math.exp(-LambdaShort * (measuredDistance + deltaR))
        return upper - lower
    # else:
    #     yita = 1.0 / (1 - math.exp(-LambdaShort * realDistance))
    #     ps = yita * LambdaShort * math.exp(-LambdaShort * measuredDistance)
    #     return ps


def GetPMax(measuredDistance, deltaR):
    if math.fabs(MaxRange - measuredDistance < deltaR):
        return 1.0 / deltaR
    else:
        return 0.0


def GetPRandom(measuredDistance, deltaR):
    if measuredDistance > MaxRange:
        return 0.0
    else:
        return deltaR / MaxRange


def Getprobability(robotToWall, particleToWall, deltaR):
    if particleToWall > MaxRange:
        particleToWall = MaxRange

    # GHit = WeighingFactors.ZHit
    # GPHit = GetPHit(robotToWall, particleToWall, deltaR)
    # 1.0 0.1 0.0 0.1
    probability = WeighingFactors.ZHit * GetPHit(robotToWall, particleToWall, deltaR) + \
                  WeighingFactors.ZShort * GetPShort(robotToWall, particleToWall, deltaR) + \
                  WeighingFactors.ZMax * GetPMax(robotToWall, deltaR) + \
                  WeighingFactors.ZRand * GetPRandom(robotToWall, deltaR)

    return probability


def SamplePHit(realDistance):
    # realDistance = int(realDistance)
    m_NormalDistribution = [realDistance, MeasurementSigma_get()]
    # m_NormalDistribution.Sigma = MeasurementSigma
    while True:
        measurementSample = np.random.normal(m_NormalDistribution[0], m_NormalDistribution[1])
        if measurementSample < MaxRange:
            break
    return measurementSample


def SamplePShort(realDistance):
    while True:
        measurementSample = np.random.exponential(realDistance)
        if measurementSample < realDistance:
            break
    return measurementSample


def SamplePRandom():
    return np.random.random() * MaxRange


def GetSample(realDistance):
    RealDistance = realDistance / 20.0
    effectiveDistance = MaxRange if realDistance > MaxRange else realDistance
    if realDistance > MaxRange:
        return MaxRange

    Random = np.random.random()

    if Random <= WeighingFactors.ZHit:
        return SamplePHit(effectiveDistance)
    elif Random <= WeighingFactors.ZHit + WeighingFactors.ZShort:
        return SamplePShort(effectiveDistance)
    elif Random <= WeighingFactors.ZHit + WeighingFactors.ZShort + WeighingFactors.ZMax:
        return MaxRange
    else:
        return SamplePRandom()


def BeamModel(maxRange, measurementVariance, lambdaShort):
    global MaxRange, LambdaShort
    MaxRange = maxRange
    setMeasurementVariance (measurementVariance)
    MeasurementSigma_set (math.sqrt(getMeasurementVariance()))
    LambdaShort = lambdaShort

File: MCL_by_Cpp/MCL_C++/BeamModel.cpp
#include "stdafx.h"
#include <iostream>
#include <cmath>
#include <ctime>
//#include <windows.system.h>
#include "BeamModel.h"
#include "CommonFunctions.h"
#include "Points.h"
#include "WeighingFactors.h"
using namespace std;

class MeasurementVariance{
private: 
  double x; 
public:  
  double getX()
  {
    return x;
  }
  	
  void setX(double value) 
  {
    x = value;
  }    
}MV;

class MeasurementSigma{
private: 
  double x; 
public:  
  double getX()
  {
    return x;
  }
  	
  void setX(double value) 
  {
    x = value;
  }    
}MS;

int MaxRange = 0;
double LambdaShort = 0;
double ZHit = 0, ZShort = 0, ZRand = 0, ZMax = 0;

void BeamModel(int maxRange, double measurementVariance, double lambdaShort, double deltar)
{
  /// Initial Beam_model parameter
  
  MaxRange = maxRange;
  MV.setX(measurementVariance);
  MS.setX(sqrt(MV.getX()));
  LambdaShort = lambdaShort;
  DeltaR = deltar;
}


double generateGaussianNoise(double mu, double sigma)
{
  /// Gaussian distribution
  
  //srand((unsigned int)time(0));
  while(1)
  {
    double u1 = double(rand() % 1000) / 1000, u2 = double(rand() % 1000) / 1000, r;
    static unsigned int seed = 0;
    r = mu + sqrt ( sigma)*sqrt(-2.0*(log(u1) / log(exp(1.0))))*cos(2 * p_PI * u2);
    if (!isinf(r) && !isnan(r))
    {
      return r;
    }
  }
};

double randomExponential(double lambda)
{
  /// Exponential distribution
  
    double pV = 0.0;
    while(true)
    {
        pV = (double)rand()/(double)RAND_MAX;
        if (pV != 1)
        {
            break;
        }
    }
    pV = (-1.0/lambda)*log(1-pV);
    return pV;
}

double GenerateRandom()
{
  //srand((unsigned int)time(0));
	
  return ((double)rand())/RAND_MAX;
}


double SamplePHit(double realDistance)
{
  /// Sample P hit
 
  double measurementSample;
  do
  {
    measurementSample = generateGaussianNoise(realDistance, MS.getX());
  }while (measurementSample > MaxRange || measurementSample < 0);
  
  return measurementSample;
}

double SamplePShort(double realDistance)
{
  /// Sample P short
  
  double measurementSample;
  do
  {
    measurementSample = randomExponential(LambdaShort);
  }while(measurementSample > realDistance);
  
  return measurementSample;
}

double SamplePRandom()
{
  /// Sample P rand
  
  return GenerateRandom() * MaxRange;
}

double GetSample(double realDistance)
{
    double effectiveDistance =  (realDistance > MaxRange) ? MaxRange : realDistance;
    if (realDistance > MaxRange)
        return MaxRange;
    
    
    /// Creates a distance measurement sample under the assumption that the real distance
    /// is the provided distance value.
    /// 
    /// A distance measurement sample.
    
    /// First we figure out what probability distribution to use
    
    
    double Random = GenerateRandom();  // Generate random number from 0 to 1

    if (Random <= ZHitRaw)
        return SamplePHit(effectiveDistance);
    else if (Random <= ZHitRaw + ZShort)
        return SamplePShort(effectiveDistance);
    else if (Random <= ZHitRaw + ZShort + ZMax)
        return MaxRange;
    else
        return SamplePRandom();
}


double GetPHit(double measuredDistance, double realDistance, double deltaR)
{
    if (realDistance > MaxRange)
        return 1 - DistributionFunction(MaxRange, MaxRange, MV.getX());
    else
    {
		double lower = DistributionFunction((measuredDistance - deltaR), realDistance, MV.getX());
		double upper = DistributionFunction((measuredDistance + deltaR), realDistance, MV.getX());
		return upper - lower;
    }
}

double GetPShort(double measuredDistance, double realDistance, double deltaR)
{
    if (measuredDistance > realDistance)
        return 0.0;
    else
    {
        double lower = 1 - exp(-LambdaShort * (measuredDistance - deltaR));
        double upper = 1 - exp(-LambdaShort * (measuredDistance + deltaR));
        return upper - lower;
    }
}

double GetPMax(double measuredDistance, double deltaR)
{
    if (fabs(MaxRange - measuredDistance < deltaR))
        return 1.0 / deltaR;
    else
        return 0.0;
}

double GetPRandom(double measuredDistance, double deltaR)
{
    if (measuredDistance > MaxRange)
        return 0.0;
    else
        return deltaR / MaxRange;
}


double Getprobability(double robotToWall, double particleToWall, double deltaR)
{
  /// Algorithm beam_range_finder_model(zt, xt, m)
  /// Probabilistic Robotics P119 

    if (particleToWall > MaxRange)
        particleToWall = MaxRange;
    
    // 1.0 0.1 0.0 0.1
    double probability = ZHit * GetPHit(robotToWall, particleToWall, deltaR) 
                         + ZShort * GetPShort(robotToWall, particleToWall, deltaR) 
			 + ZMax * GetPMax(robotToWall, deltaR) 
		         + ZRand * GetPRandom(robotToWall, deltaR);

    return probability;
}